### PR TITLE
(feat)  O3-3250:  Add in-patient notes workspace

### PR DESCRIPTION
**Pull request by usamaidrsk**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary

This PR adds in-patient notes workspace

## Screenshots

image

https://github.com/user-attachments/assets/78ce5648-6953-415e-b74f-1ae4aca30702

## Other

[O3-3250](https://openmrs.atlassian.net/browse/O3-3250)

[O3-3250]: https://openmrs.atlassian.net/browse/O3-3250?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by mogoodrich — COMMENTED** [review]
This generally looks good to me, but I am going to defer to @chibongho  and @brandones  who has been working with this code longer.

One thing I'm missing... where is the code that handles saving the notes?

Also fyi looks like there are some failing tests.

**Comment by usamaidrsk** [comment]
> One thing I'm missing... where is the code that handles saving the notes?

This is handled in the `esm-patient-notes-app` and the work is done in this [PR](https://github.com/openmrs/openmrs-esm-patient-management/pull/1218). For the failing tests there is some parallel work needed on the `esm-core` too, that is work in progress.
cc @mogoodrich

**Comment by mogoodrich** [comment]
> This is handled in the `esm-patient-notes-app` and the work is done in this [PR](https://github.com/openmrs/openmrs-esm-patient-management/pull/1218). For the failing tests there is some parallel work in progress on the `esm-core` too, that I am completing.

Ah, okay, so basically this rendered the existing visit note within the Ward context?

**Comment by usamaidrsk** [comment]
> Ah, okay, so basically this rendered the existing visit note within the Ward context?

I created a separate form for the ward in-patient note form.

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx
Comment (mogoodrich) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx:16]: this is the functionality that updates with bed is active/highlighted?  I would call it updateSelectedBed or updateActiveBed... "activeBedSelection" is a bit redundant and confusing.

Also, just to clarify, what happens in the cases where their are two patients in a bed?

Comment (usamaidrsk) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx:16]: > what happens in the cases where their are two patients in a bed?
The workspace opens for a single patient. So the expected is that when a patient card is selected it sets the store with the `patient`, `bed` and `visit` details.

> I would call it updateSelectedBed or updateActiveBed... "activeBedSelection" is a bit redundant and confusing.
On this let me work on a relative variable name

Thanks @mogoodrich

File: __mocks__/patient-note-config.mock.ts
Comment (mogoodrich) [inline __mocks__/patient-note-config.mock.ts:2]: Oh, sorry @usamaidrsk  I should have noted that when moving this we can now rely on the EMR API properties configuration endpoint to fetch these values:

/openmrs/ws/rest/v1/emrapi/configuration

The constants you've laid out below should map to the following properties in the configuration REST response

"clinicanEncounterRole" -> "clinicianEncounterRole.uuid" 
"encounterTypeUuid" -> "visitNoteEncounterType.uuid"
"encounterNoxtTextConceptUuid" -> "consultFreeTextCommentsConcept.uuid"

And I don't think we need the other three properties (at least not yet)
"visitDiagnosesConceptUuid"
"problemLIstConceptUuid"
"formConceptUuid"

File: packages/esm-ward-app/src/config-schema.ts
Comment (mogoodrich) [inline packages/esm-ward-app/src/config-schema.ts:34]: We shouldn't need this at, hopefuly, see my comment above

File: packages/esm-ward-app/src/types/index.ts
Comment (mogoodrich) [inline packages/esm-ward-app/src/types/index.ts:167]: This can just be called "EncounterPayload".... ideally it would be a type already defined in ESM Core somewhere, but if not, fine to define it here. (fyi @ibacher @denniskigen )

File: packages/esm-ward-app/src/types/index.ts
Comment (mogoodrich) [inline packages/esm-ward-app/src/types/index.ts:179]: Is there an existing type defined in ESM Core we can use instead, otherwise fine to define it here. (fyi @ibacher @denniskigen )

File: packages/esm-ward-app/src/ward-workspace/patient-banner/patient-banner.component.tsx
Comment (mogoodrich) [inline packages/esm-ward-app/src/ward-workspace/patient-banner/patient-banner.component.tsx:9]: Will defer to @chibongho  when he gets back on this part...

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes-config-schema.ts
Comment (mogoodrich) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes-config-schema.ts:1]: Hopefully we can get rid of all this per my comments above.

**Review by mogoodrich — CHANGES_REQUESTED** [review]
Generally looks great @usamaidrsk , but we should fetch the configuration from the "configuration" endpoint instead of setting up a new notes config, see my comments... sorry, I should have mentioned that before, didn't think of it.

I also flagged Ian and Dennis for a few years, but this is close, thanks, I'm excited to see this working!

**Comment by mogoodrich** [comment]
Oh, and also, looks like some merge errors...

File: packages/esm-ward-app/src/ward-workspace/patient-banner/patient-banner.component.tsx
Comment (usamaidrsk) [reply packages/esm-ward-app/src/ward-workspace/patient-banner/patient-banner.component.tsx:9]: @chibongho I hope you can now take a look at this. Thanks

File: packages/esm-ward-app/src/types/index.ts
Comment (mogoodrich) [reply packages/esm-ward-app/src/types/index.ts:168]: let's just go ahead then and rename this EncounterPayload

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx
Comment (mogoodrich) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx:18]: is this meant to be commented out?

**Review by mogoodrich — APPROVED** [review]
@usamaidrsk  I made a few small requests for changes, but generally LGTM, I approved.
Would be still be good to get a review from @chibongho   Thanks!

File: __mocks__/emr-api-config.mock.ts
Comment (vasharma05) [inline __mocks__/emr-api-config.mock.ts:1]: Put this mock inside the `ward-app` mocks inside the __mocks__ folder.

File: packages/esm-ward-app/src/index.ts
Comment (vasharma05) [inline packages/esm-ward-app/src/index.ts:47]: Why are we fetching these synchronously?

Extensions which should render directly on the UI on the first load, say Ward View, should be imported synchronously to reduce the API calls to fetch data, else all other things should be imported asynchronously.

Comment (vasharma05) [reply packages/esm-ward-app/src/index.ts:47]: Few examples will be:
1. Left nav items on Home page/ Patient chart
2. Items on the Primary Navigation

File: packages/esm-ward-app/src/routes.json
Comment (vasharma05) [inline packages/esm-ward-app/src/routes.json:46]:
```suggestion
      "name": "ward-inpatient-notes-form-action-button",
```

File: packages/esm-ward-app/src/routes.json
Comment (vasharma05) [inline packages/esm-ward-app/src/routes.json:63]: Why are we using `width` as wider?
It should be `narrow`, which is the default value.

Only clinical forms have `wider` width and all other workspaces (except ward patient workspace) use the `narrow` width for the workspaces.

File: packages/esm-ward-app/src/types/index.ts
Comment (vasharma05) [inline packages/esm-ward-app/src/types/index.ts:8]:
```suggestion
import type {
  Concept,
  Location,
  OpenmrsResource,
  OpenmrsResourceStrict,
  Patient,
  Person,
  Visit,
```

File: packages/esm-ward-app/src/types/index.ts
Comment (vasharma05) [inline packages/esm-ward-app/src/types/index.ts:177]: You can remove the comments in the interface, as most developers are/should be thorough with core structures (Encounters, Visits, etc...)

Also, a tip, if you want to specify a property, use [JSDocs](https://jsdoc.app/about-getting-started) for the same. Eg.

```
/**
 * date and time the encounter was created (ISO8601 Long) 
 */
encounterDatetime: string; 
```

File: packages/esm-ward-app/src/types/index.ts
Comment (vasharma05) [reply packages/esm-ward-app/src/types/index.ts:180]: No @mogoodrich , I didn't find it,  but it should be present in the core.

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx:20]: Passing the type inside the `launchWorkspace` helps us to correctly determine the data to pass into the function, which is expected on the other side by the workspace.
Please update the type accordingly and add this back.
Thanks!

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card.tsx:17]: If we don't have any conditions to take care in the classes, we can ignore using classnames.

File: packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx:19]: In case someone doesn't pass the `patient` inside the workspace, `patient` will be undefined and this will surely break the code, which should be handled.
Please try not to write something, which might break.
Thanks!

File: packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx:21]: What are we getting from the `patient` prop, that we are re-fetching the patient here?
Thanks!

File: packages/esm-ward-app/src/ward-workspace/patient-banner/patient-banner.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/patient-banner/patient-banner.component.tsx:10]:
```suggestion
  const { location } = useWardLocation();
```

File: packages/esm-ward-app/src/ward-workspace/patient-banner/style.scss
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/patient-banner/style.scss:1]:
```suggestion
@use '@carbon/layout';
```

File: packages/esm-ward-app/src/ward-workspace/patient-banner/style.scss
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/patient-banner/style.scss:9]:
```suggestion
  padding: layout.$spacing-04;
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:9]:
```suggestion
import { type DefaultWorkspaceProps } from '@openmrs/esm-framework';
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:24]:
```suggestion
interface PatientNotesFormProps extends DefaultWorkspaceProps {
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:57]: Function should be wrapped in `useCallback`

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:48]:
```suggestion
  const { control, handleSubmit, watch, getValues, setValue, formState: {isDirty} } = useForm<NotesFormData>({
    mode: 'onSubmit',
    resolver: zodResolver(noteFormSchema),
    defaultValues: {
      wardClinicalNote: '',
    },
  });
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:109]: On error, we should display all the errors as inline notifications at the end of the form.
The `errors` is not of instance `Error` but an object containing the field name as key and **error messages** as values.

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:166]: Don't we have a Cancel button in the workspace 🤔

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.resource.ts
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.resource.ts:4]:
```suggestion
export function savePatientNote(payload: EncounterPayload, abortController: AbortController = new AbortController()) {
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx:30]: As per updates from @denniskigen, we no longer need to mock the @openmrs/esm-framework.
```
const mockedcreateErrorHandler = jest.mocked (createErrorHandler)
const mockedshowSnackbar = jest.mocked (  showSnackbar)
const mockedtranslateFrom = jest.mocked (  translateFrom)
const mockedResponsiveWrapper = jest.mocked (  ResponsiveWrapper)
const mockeduseSession = jest.mocked (  useSession).mockReturnValue(() => mockSession)

```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx:39]:
```suggestion
jest.mock('../../../hooks/useEmrConfiguration', () => jest.fn() );
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx:41]:
```suggestion
const mockedUseEmrConfiguration = jest.mocked(useEmrConfiguration)
mockedEmrConfiguration.mockReturnValue({
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes-action-button.extension.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes-action-button.extension.tsx:5]: We are not getting any props in the Action Button.
Please remove this as this will pass unnecessary props inside the `launchWorkspace` function.
Thanks!

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes-action-button.extension.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes-action-button.extension.tsx:13]:
```suggestion
      handler={() => launchWorkspace('ward-patient-notes-workspace')}
```

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes.style.scss
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes.style.scss:6]: The width issue needs to be fixed in the workspace container in the core. Please don't add these stylings here, else after the fix is made there, it will break here.

**Review by vasharma05 — CHANGES_REQUESTED** [review]

**Comment by vasharma05** [comment]
Hi @usamaidrsk!
I have some suggestions for improving the code structure before merging the PR.

@mogoodrich, I don't mean to overcome your review, I have some suggestions which should go in before the changes are merged.
Thanks!

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes.style.scss
Comment (brandones) [reply packages/esm-ward-app/src/ward-workspace/ward-patient-notes/notes.style.scss:6]: Also the height styling isn't necessary. If anything is needed for height, it will be `height: 100%`.

**Comment by mogoodrich** [comment]
> @mogoodrich, I have some suggestions which should go in before the changes are merged.
> Thanks!

Thanks @vasharma05 !  Defer to you when this is ready to be merged.

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (usamaidrsk) [reply packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:109]: These are validation errors, that are handled by the `invalid` and `invalidText` attributes on the form fields

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (usamaidrsk) [reply packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:166]: The design does not include one on this

File: packages/esm-ward-app/src/index.ts
Comment (vasharma05) [inline packages/esm-ward-app/src/index.ts:47]: Can you please make this async too?
Thanks!

File: packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx:18]: It's better to export types from a separate file, else, sometimes, the components gets circular imports due to type importing.

Comment (vasharma05) [reply packages/esm-ward-app/src/ward-patient-workspace/ward-patient.workspace.tsx:18]: If a sub component needs this type, it will import the type from the parent and parent imports sub component, hence cyclic imports can happen.

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:71]: Don't send encounterDatetime from the server. It is automatically added by the BE server. This will otherwise put us into issues like this: https://github.com/openmrs/openmrs-ngx-formentry/pull/131

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.component.tsx:80]: This is also handled by the BE itself. It'll assign the role correctly given the auth token. 
Don't send this to the Backend.

File: packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx
Comment (vasharma05) [inline packages/esm-ward-app/src/ward-workspace/ward-patient-notes/form/notes-form.test.tsx:68]:
```suggestion
  mockSavePatientNote.mockResolvedValue({ status: 201, body: 'Condition created' });
```

In case you might have missed, ignore if intentionally done.

**Review by vasharma05 — COMMENTED** [review]
A few more changes. Great work @usamaidrsk!


One thing I'm wondering, the notes form is in the same app as of the Extension slot where it is being rendered, why aren't we directly importing it into the workspace?
Cc: @mogoodrich @chibongho

File: packages/esm-ward-app/src/index.ts
Comment (vasharma05) [inline packages/esm-ward-app/src/index.ts:42]:
```suggestion
// t("inpatientNotesWorkspaceTitle", "In-patient notes")
export const wardPatientNotesWorkspace = getAsyncLifecycle(
```

Comment (vasharma05) [reply packages/esm-ward-app/src/index.ts:42]: Translation key for the workspace title needs to be added in the `en.json`

**Review by vasharma05 — APPROVED** [review]
LGTM, thanks @usamaidrsk!

File: packages/esm-ward-app/src/index.ts
Comment (usamaidrsk) [reply packages/esm-ward-app/src/index.ts:42]: This change is added @mogoodrich, Thanks

**Comment by mogoodrich** [comment]
Thanks @usamaidrsk !  Just waiting for all the checks to pass, then I will merge.

**Pull request merged**
